### PR TITLE
PCI-209-modify-validators Complete validators manage process / subscriptions, alter tests.

### PR DESCRIPTION
**Pull request by maryjanyes**

### CODE REVIEW

File: cosmos-abci/abci/src/grpc/protos.rs [hunk 1]
```diff
@@ -164,7 +164,14 @@ impl crate::ResponseInitChain for ResponseInitChain {}
...
 impl crate::ResponseBeginBlock for ResponseBeginBlock {}
 
-impl crate::ResponseEndBlock for ResponseEndBlock {}
+impl crate::ResponseEndBlock for ResponseEndBlock {
+    fn get_validator_updates(&self) -> Vec<ValidatorUpdate> {
+        self.validator_updates.clone()
+    }
+    fn get_events(&self) -> Vec<Event> {
+        self.events.clone()
+    }
```
Comment (Mr-Leshiy) [inline cosmos-abci/abci/src/grpc/protos.rs:173]: Add also setters for these fields

```Rust
fn set_events(&self, v : Vec<Event>) {
...
}

fn set_validator_updates(&self, v :Vec<ValidatorUpdate>) {
...
}
```

File: cosmos-abci/src/lib.rs [hunk 2]
```diff
@@ -316,6 +422,11 @@ pub trait AbciInterface {
             .map_err(|_| "failed to setup connection")?
             .end_block(height)
             .map_err(|_| "end_block failed")?;
+        // debug::info!("Result: {:?}", result);
+        let cosmos_node_validators = _result.get_validator_updates();
+        debug::info!("Cosmos validators {:?}", cosmos_node_validators);
+        // todo
```
Comment (Mr-Leshiy) [inline cosmos-abci/src/lib.rs:428]:
```suggestion
        // TODO : Save cosmos node validators into storage
```

File: cosmos-abci/src/lib.rs [hunk 3]
```diff
@@ -316,6 +422,11 @@ pub trait AbciInterface {
...
+        // Save cosmos node validators into storage.
```
Comment (Mr-Leshiy) [inline cosmos-abci/src/lib.rs:429]:
```suggestion
```

File: cosmos-abci/src/lib.rs [hunk 4]
```diff
@@ -334,3 +445,111 @@ impl<T: Trait> sp_runtime::offchain::storage_lock::BlockNumberProvider for Modul
...
+        // todo
+        // Get cosmos accounts & active validators from rocks_db storage.
+        let last_cosmos_validators: Vec<utils::CosmosAccountId> = vec![];
+        let mut new_substrate_validators: Vec<T::AccountId> = vec![];
+        for cosmos_validator_id in &last_cosmos_validators {
+            let substrate_account_id = <CosmosAccounts<T>>::get(&cosmos_validator_id);
+            if substrate_account_id.is_some() {
+                if let Some(full_substrate_account_id) = substrate_account_id {
+                    new_substrate_validators.push(full_substrate_account_id);
+                }
```
Comment (Mr-Leshiy) [inline cosmos-abci/src/lib.rs:488]: Print WARNING message if we can`t find substrate account for the corresponding cosmos account

**Review by Mr-Leshiy — CHANGES_REQUESTED** [review]

**Review by Mr-Leshiy — APPROVED** [review]

**Review by enfipy — APPROVED** [review]

**Pull request merged**